core/src/main/java/org/apache/iceberg/ManifestExpirationManager.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.commons.compress.utils.Sets;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ManifestExpirationManager {
+
+  //No public constructor for utility class
+  private ManifestExpirationManager(){}
+
+  private static final Logger LOG = LoggerFactory.getLogger(ManifestExpirationManager.class);
+
+  /**
+   * Scans a set of manifest files for deleted data files which do not refer to a valid snapshot and can be actually
+   * removed from the filesystem. This implementation will use a concurrent approach and will ignore any
+   * failed manifest reads.
+   *
+   * @param manifests        Manifests to scan, all of these files will be scanned
+   * @param validSnapshotIds Snapshot Ids currently considered valid in Table metadata
+   * @param specLookup       A mapping between partitionID and the spec describing that partition
+   * @param io               IO used for reading the files
+   * @return The set of all files that can be safely deleted
+   */
+  public static Set<String> scanManifestsForAbandonedDeletedFiles(
+      Set<ManifestFile> manifests, Set<Long> validSnapshotIds, Map<Integer, PartitionSpec> specLookup,
+      FileIO io) {
```

Comment:
Since we are introducing a class for this, I think it makes sense to simplify the methods. Before, we kept breaking out new methods to keep complexity down, which kept requiring passing all of the information as arguments. Now, we can share many of these arguments, like `validSnapshotIds` in the manager instance and not need to pass them around as args.

We can also initialize the manager using the base snapshot and current metadata, so that the class handles its own configuration for things like `specLookup`.

What do you think?

Reply:
As long as I can keep all the fields final and serializable I'm ok with that. I have too much Scala in my brain and I generally hate having any methods on a class if I can help it and prefer to just have data classes and then static methods which operate on them.

 I think in this case all the inputs to this class are also serializable so it's probably safe ...

Reply:
This ends up being a bit more complicated than I anticipated if we want to make sure that "ManfiestExpirationChanges" and the Manager are both not only serializable but also splittable ... I think if I do a bit more refactoring I can make this happen but I believe this will make it harder/ more confusing in the future. 

For example I'll need this class to not only generate the full list of manifests which need to be scanned but also have a method which only operates on some of the generated list.

Also we end up loosing a bit of control about how we serialize the component parts, we would have always broadcast everything if we wrap the entire module up.

I can go ahead and finish up the refactor but I would like to check with you before I devote more time to this.

My code at the moment is a bit like

```
//These are all serializable
  private final SnapshotExpirationChanges snapshotChanges; 
  private final ManifestExpirationChanges manifestExpirationChanges;
  private final Map<Integer, PartitionSpec> specLookup;
  private final FileIO io;

  public ManifestExpirationManager(TableMetadata original, TableMetadata current, FileIO io){
    this.io = io;
    this.specLookup = current.specsById();
    this.snapshotChanges = ExpireSnapshotUtil.getExpiredSnapshots(current, original);
    this.manifestExpirationChanges = ExpireSnapshotUtil.determineManifestChangesFromSnapshotExpiration(
        snapshotChanges.getValidSnapshotIds(), snapshotChanges.getExpiredSnapshotIds(), current, original, io)
    )
  //We can simplify the functions above and move them into this class

// But then we need to be able to get manifestExpirationChanges because we want to parallelize information from them
  getManifestExpirationChanges()

 // And now our scan method has to take a list of manifests still or we have to pass some kind of partition/split info
  public Set<String> scanManifestsForAbandonedDeletedFiles(Set<ManifestFile> manifestsWithDeletes) {

  }
```

Let me know if this is kind of what you were thinking about

Reply:
I pushed a commit with this as a work in progress. I tried to keep the class as stateless as possible but it will have some null handling behaviors to maintain old effects.

Reply:
See https://github.com/apache/iceberg/pull/1264 For a demonstration of how we would use this in Spark